library/choose_disk.py

```diff
@@ -0,0 +1,479 @@
+        device_name = os.readlink(directory + f).split("/")[-1]
+        if device_name in physical_disks:
+            if device_name not in matching_devices:
+                matching_devices[device_name] = [f]
+            else:
+                matching_devices[device_name].append(f)
+
+    for physical_disk in sorted(physical_disks):
+        if physical_disk in matching_devices:
+            current_index = sorted(matching_devices[physical_disk])[0]
+            final_disks[current_index] = physical_disks[physical_disk]
+            final_disks[current_index]["bdev"] = "%s%s" % (directory, current_index)
+            info(' Renaming %10s to %50s' % (physical_disk, current_index))
+        else:
+            current_index = physical_disk
+            final_disks[current_index] = physical_disks[physical_disk]
+
+    return final_disks
+
+
+def fake_device(legacy_devices, ceph_type):
+    '''
+    In case of legacy block device names, let's create an internal faked
+    entry with a 'bdev' entry filled with the actual path. This will be used to
+    make a match later on.
+    '''
+    devices = {}
+    count = 0
+    for device in legacy_devices:
+        devices["%s_%d" % (ceph_type, count)] = {"bdev": os.path.dirname(device)+"/"+os.path.basename(device)}
+        count = count + 1
+
+    return devices
+
+
+def show_resulting_devices(matched_devices, physical_disks):
+    bdev_matched = []
+    bdev_unmatched = []
+    info("Matched devices   : %3d" % len(matched_devices))
+    for matched_device in sorted(matched_devices):
+        extra_string = ""
+        if "ceph_prepared" in matched_devices[matched_device]:
+            extra_string = " (ceph)"
+        info(" %s : %s%s" % (matched_device, matched_devices[matched_device]["bdev"], extra_string))
+        bdev_matched.append(matched_devices[matched_device]["bdev"])
+
+    for physical_disk in sorted(physical_disks):
+        if physical_disks[physical_disk]["bdev"] not in bdev_matched:
+            bdev_unmatched.append(physical_disks[physical_disk]["bdev"])
+
+    info("Unmatched devices : %3d" % len(bdev_unmatched))
+    for bdev in sorted(bdev_unmatched):
+        info(" %s" % bdev)
+
+
+def info(message):
+    global logger
+    if logger:
+        logger.info(message)
+
+
+def warn(message):
+    global logger
+    if logger:
+        logger.warn(message)
+
+
+def debug(message):
+    global logger
+    if logger:
+        logger.debug(message)
+
+
+def error(message):
+    global logger
+    if logger:
+        logger.error(message)
+
+
+def setup_logging():
+    global logger
+    hdlr = logging.FileHandler('/var/log/choose_disk.log')
+    formatter = logging.Formatter('%(asctime)s %(levelname)s %(message)s')
+    hdlr.setFormatter(formatter)
+    logger.addHandler(hdlr)
+    logger.setLevel(logging.INFO)
+    logger.info("############")
+    logger.info("# Starting #")
+    logger.info("############")
+
+
+def fatal(message):
+    global module
```

Comment:
There are lots of issues of using `global` and I am not seeing why it is necessary here (other than to avoid passing it into functions).

Using globals in this way will obfuscate where things are coming from, making it harder to debug when/if something fails. When there are side-effects it also makes it hard to determine what/where is causing them.

Refactoring gets affected as well, because the scope becomes unclear. The namespace gets polluted, and there is potential for overwriting or redefining the global unknowingly. 

Linters will have a hard time detecting some issues associated with assignment too.

A global shouldn't be used just because it avoids the inconvenience of passing a variable around.